pkg/trace/api/otlp.go

```diff
@@ -203,70 +203,136 @@ func fastHeaderGet(h http.Header, canonicalKey string) string {
 func (o *OTLPReceiver) processRequest(protocol string, header http.Header, in ptraceotlp.Request) {
 	for i := 0; i < in.Traces().ResourceSpans().Len(); i++ {
 		rspans := in.Traces().ResourceSpans().At(i)
-		// each rspans is coming from a different resource and should be considered
-		// a separate payload; typically there is only one item in this slice
-		attr := rspans.Resource().Attributes()
-		rattr := make(map[string]string, attr.Len())
-		attr.Range(func(k string, v pcommon.Value) bool {
-			rattr[k] = v.AsString()
-			return true
-		})
-		lang := rattr[string(semconv.AttributeTelemetrySDKLanguage)]
-		if lang == "" {
-			lang = fastHeaderGet(header, headerLang)
-		}
-		tagstats := &info.TagStats{
-			Tags: info.Tags{
-				Lang:            lang,
-				LangVersion:     fastHeaderGet(header, headerLangVersion),
-				Interpreter:     fastHeaderGet(header, headerLangInterpreter),
-				LangVendor:      fastHeaderGet(header, headerLangInterpreterVendor),
-				TracerVersion:   fmt.Sprintf("otlp-%s", rattr[string(semconv.AttributeTelemetrySDKVersion)]),
-				EndpointVersion: fmt.Sprintf("opentelemetry_%s_v1", protocol),
-			},
-			Stats: info.NewStats(),
-		}
-		tracesByID := make(map[uint64]pb.Trace)
-		for i := 0; i < rspans.ScopeSpans().Len(); i++ {
-			libspans := rspans.ScopeSpans().At(i)
-			lib := libspans.Scope()
-			for i := 0; i < libspans.Spans().Len(); i++ {
-				span := libspans.Spans().At(i)
-				traceID := traceIDToUint64(span.TraceID().Bytes())
-				if tracesByID[traceID] == nil {
-					tracesByID[traceID] = pb.Trace{}
+		o.ReceiveResourceSpans(rspans, header, protocol)
+	}
+}
+
+// OTLPIngestSummary returns a summary of the received resource spans.
+type OTLPIngestSummary struct {
+	// Hostname indicates the hostname of the passed resource spans.
+	Hostname string
+	// Tags returns a set of Datadog-specific tags which are relevant for identifying
+	// the source of the passed resource spans.
+	Tags []string
+}
+
+// ReceiveResourceSpans processes the given rspans and sends them to writer.
+func (o *OTLPReceiver) ReceiveResourceSpans(rspans ptrace.ResourceSpans, header http.Header, protocol string) OTLPIngestSummary {
+	// each rspans is coming from a different resource and should be considered
+	// a separate payload; typically there is only one item in this slice
+	attr := rspans.Resource().Attributes()
+	rattr := make(map[string]string, attr.Len())
+	attr.Range(func(k string, v pcommon.Value) bool {
+		rattr[k] = v.AsString()
+		return true
+	})
+	hostname, _ := attributes.HostnameFromAttributes(attr)
+	if hostname == "" {
+		hostname = rattr["_dd.hostname"]
+	}
```

Comment:
I think we should check the second value returned by this function, since in some environments (e.g. ECS Fargate) an empty hostname is actually valid and what we want to use (or at least this is how it works in metrics/host metadata 😄).

```suggestion
	hostname, ok := attributes.HostnameFromAttributes(attr)
	if !ok {
		hostname = rattr["_dd.hostname"]
	}
```

Reply:
Very good catch. Thank you for this. Changed. It is exactly the same in APM. The regular trace-agent pipeline does detect the Fargate orchestrator but [that functionality](https://github.com/DataDog/datadog-agent/blob/aca042fe8bf129c856bbedb9f38514dcf881e7fa/cmd/trace-agent/config/config.go#L69) is not available here so this is a good change.